validation: docs and cleanups for MemPoolAccept coins views

glozow opened the pull request:
Deletes `test_mid_package_eviction` that is brittle and already covered in other places. It was introduced in #28251 addressing 2 issues: (1) calling `LimitMempoolSize()` in the middle of package validation and (2) not updating coins view cache when the mempool contents change, leading to "disappearing coins."

(1) If you let `AcceptSingleTransaction` call `LimitMempoolSize` in the middle of package validation, you should get a failure in `test_mid_package_eviction_success` (the package is rejected):
```
diff --git a/src/validation.cpp b/src/validation.cpp
index f2f6098e214..4bd6f059849 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1485,7 +1485,7 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
     FinalizeSubpackage(args);
 
     // Limit the mempool, if appropriate.
-    if (!args.m_package_submission && !args.m_bypass_limits) {
+    if (!args.m_bypass_limits) {
         LimitMempoolSize(m_pool, m_active_chainstate.CoinsTip());
         // If mempool contents change, then the m_view cache is dirty. Given this isn't a package
         // submission, we won't be using the cache anymore, but clear it anyway for clarity.
```
Mempool modifications have a pretty narrow interface since #31122 and `TrimToSize()` cannot be called while there is an outstanding mempool changeset. So I think there is a low likelihood of accidentally reintroducing this problem and not immediately hitting e.g. a fuzzer crash on this line https://github.com/bitcoin/bitcoin/blob/b53fab1467fde73c40402e2022b25edfff1e4668/src/txmempool.cpp#L1143

(2) If you remove the `CleanupTemporaryCoins()` call from `ClearSubPackageState()` you should get a failure from `test_mid_package_replacement`:
```
diff --git a/src/validation.cpp b/src/validation.cpp
index f2f6098e214..01b904b69ef 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -779,7 +779,7 @@ private:
         m_subpackage = SubPackageState{};
 
         // And clean coins while at it
-        CleanupTemporaryCoins();
+        // CleanupTemporaryCoins();
     }
 };
```
I also added/cleaned up the documentation about coins views to hopefully make it extremely clear when people should `CleanupTemporaryCoins`.

glozow committed:
[doc] MemPoolAccept coins views

glozow committed:
don't make a copy of m_non_base_coins

glozow committed:
[doc] remove references to now-nonexistent Finalize() function

glozow committed:
[cleanup] delete brittle test_mid_package_eviction

This test was introduced in #28251 to ensure that the mempool is not
trimmed in the middle of a package evaluation and the m_view cache
is updated when evictions and replacements happen so coins are no longer
visible in subsequent package transactions. These two things have
coverage in other tests as well, and are pretty unlikely to happen.

This test is also brittle: it requires evaluation of the parents in a
particular order, and creates a transaction that itself is not
enough to trigger eviction but will be pushed out immediately by the
package spending from it. While the current magic number 2000 works, we
do not have a way to query remaining space in the mempool if mempool
data structures change, and it can differ across platforms.

glozow committed:
[doc] always CleanupTemporaryCoins after a mempool trim

instagibbs commented:
in op and commit message: `test_mid_package_evaluation` doesn't exist, it's named `test_mid_package_eviction`

The bull case for keeping `test_mid_package_replacement` is that it involves an RBF, not a mempool overflow, right? The comment in that test seems slightly erroneous, mentioning `replaced_tx` needs to be at the bottom of the mempool, when I think it just has to be low feerate enough to be RBF'd?

instagibbs commented on src/validation.cpp:
```diff
@@ -725,8 +725,27 @@ class MemPoolAccept
     CTxMemPool& m_pool;
+
+    /** Holds a cached view of available coins from the UTXO set, mempool, and artificial temporary coins (to enable package validation).
+     * The view doesn't track whether a coin previously existed but has now been spent. We detect conflicts in other ways:
```
> The view doesn't track whether a coin previously existed but has now been spent

been spent in mempool/package contexts, I presume

instagibbs replied:
s/has now been spent/has now been spent in the mempool/ ?

glozow replied:
Will change if I need to retouch

instagibbs reviewed (approved):
ACK b5805eddefced4cd8bb6f8069ea0a4bf199bacff

confirmed that the remaining tests cases fail when specific key clauses are modified, that it seems implausible that the test adds much assurances, and honestly is not a particularly maintainable test. The new comments are helpful for me to trace the code in a directed way and confirm what it's saying.

glozow committed:
[doc] reword comments in test_mid_package_replacement

The comment about eviction seems to be erroneously copy-pasted. Reword
another comment for clarity.

glozow commented:
> op and commit message: `test_mid_package_evaluation` doesn't exist, it's named `test_mid_package_eviction`

Fixed

> The bull case for keeping `test_mid_package_replacement` is that it involves an RBF, not a mempool overflow, right? The comment in that test seems slightly erroneous, mentioning `replaced_tx` needs to be at the bottom of the mempool, when I think it just has to be low feerate enough to be RBF'd?

Thanks, yeah, I think it had an incorrect copy-paste from the other test. I think this one is good to keep for when we're working on generalized package validation.

instagibbs commented:
reACK https://github.com/bitcoin/bitcoin/pull/32973/commits/b6d4688f77df9e31fd64e2be300f55bb8e944bd0

marcofleon reviewed (commented):
ACK b6d4688f77df9e31fd64e2be300f55bb8e944bd0

Verified that the two functional tests fail with the suggested buggy patches. Seems fine to remove this test if it doesn't provide any new assurances.

> I think there is a low likelihood of accidentally reintroducing this problem and not immediately hitting e.g. a fuzzer crash on this line

I also tried to see if we'd get a fuzz crash after applying the first patch and it doesn't seem to be happening. I'm running `tx_package_eval` with a small mempool in order to trigger the assertion in `TrimToSize()`. I also commented out the added `CleanupTemporaryCoins()` and still nothing. Fwiw if I add `LimitMempoolSize()` right before `ClearSubPackageState()` in `AcceptSubPackage` there is an almost immediate crash.

If I'm misunderstanding the package validation code or running the wrong target, let me know. I haven't looked into this too thoroughly (e.g. checking coverage) but I plan on investigating some more later.

glozow commented:
>  Fwiw if I add LimitMempoolSize() right before ClearSubPackageState() in AcceptSubPackage there is an almost immediate crash.

That's the kind of scenario I was thinking - somehow reintroducing a trim somewhere mid-subpackage.

marcofleon commented:
Thanks, was able to reproduce.

```c
fuzz: ../../src/validation.cpp:428: bool CheckInputsFromMempoolAndCache(const CTransaction &, TxValidationState &, const CCoinsViewCache &, const CTxMemPool &, unsigned int, PrecomputedTransactionData &, CCoinsViewCache &, ValidationCache &): Assertion `!coinFromUTXOSet.IsSpent()' failed.
```

Now the comment for `m_view` in 98ba2b1db2eb81e08b550ba0d5069a9289f30e13 makes more sense to me as well.
